File: 25.ReverseNodeInKGroups/Solution.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution
{
    ListNode *reverseKGroup(ListNode *head, int k)
    {
        if (k == 1 or !head or !head->next)
            return head;
        ListNode *p = head;
        int len = 0;
        for (int i = 0; i < k; i++)
        {
            if (!p)
                return head;
            p = p->next;
        }
        ListNode *q = nullptr, *r = nullptr, *nextHead = p;
        p = head;
        for (int i = 0; i < k; i++)
        {
            r = q;
            q = p;
            p = p->next;
            q->next = r;
        }
        head->next = reverseKGroup(nextHead, k);
        return q;
    }
};
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

File: 25.ReverseNodeInKGroups/Solution.java
public class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        if (head == null || k == 1)
            return head;
        ListNode p = head;
        for (int i = 0; i < k; i++) {
            if (p == null)
                return head;
            p = p.next;
        }
        ListNode r = null, q = null, nextHead = p;
        p = head;
        for (int i = 0; i < k; i++) {
            r = q;
            q = p;
            p = p.next;
            q.next = r;
        }
        head.next = reverseKGroup(nextHead, k);
        return q;
    }
}

class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}
